crates/tabby-scheduler/src/index.rs

```diff
@@ -25,43 +32,44 @@ pub fn index_repositories(_config: &[RepositoryConfig]) {
     let mut writer = index
         .writer(150_000_000)
         .expect("Failed to create index writer");
+
+    let intelligence = CodeIntelligence::default();
+
     writer
         .delete_all_documents()
         .expect("Failed to delete all documents");
 
-    let total_file_size: usize = SourceFile::all()
-        .filter(is_valid_file)
-        .map(|x| x.read_file_size())
-        .sum();
-
-    let mut pb = std::io::stdout()
-        .is_terminal()
-        .then(|| tqdm(total_file_size));
-
-    let intelligence = CodeIntelligence::default();
-    for file in SourceFile::all().filter(is_valid_file) {
-        let text = match file.read_content() {
-            Ok(content) => content,
-            Err(e) => {
-                warn!("Failed to read content of '{}': {}", file.filepath, e);
+    for repository in config {
+        let Some(commit) = cache.get_last_index_commit(repository) else {
+            index_repository_from_scratch(repository, &writer, &code, &intelligence, &mut cache);
+            continue;
+        };
+        let dir = repository.dir();
+        let changed_files = get_changed_files(&dir, &commit).expect("Failed read file diff");
+        for file in changed_files {
+            let path = dir.join(&file);
+            delete_indexed_source_file(&writer, &code, &repository.git_url, &file);
+            if !path.exists() {
                 continue;
             }
-        };
+            let Some(source_file) = cache.get_source_file(repository, &path) else {
+                continue;
+            };
+            if !is_valid_file(&source_file) {
+                continue;
+            }
+            add_indexed_source_file(&writer, &source_file, &code, &intelligence);
+        }
+        cache.set_last_index_commit(
+            repository,
+            Some(get_current_commit_hash(&dir).expect("Failed to read current commit hash")),
+        );
+    }
 
-        for body in intelligence.chunks(&text) {
-            pb.as_mut()
-                .map(|b| b.update(body.len()))
-                .transpose()
-                .expect("Failed to update progress bar");
-
-            writer
-                .add_document(doc!(
-                        code.field_git_url => file.git_url.clone(),
-                        code.field_filepath => file.filepath.clone(),
-                        code.field_language => file.language.clone(),
-                        code.field_body => body,
-                ))
-                .expect("Failed to add document");
+    for indexed_repository in cache.list_indexed_repositories() {
```

Comment:
read from RepositoryAccess, instead of cache.

Reply:
This is to clean up repositories which have been deleted but are still present in the index, reading from RepositoryAccess won't solve it

Reply:
it can be handled with garbage collection: https://github.com/TabbyML/tabby/pull/2063